### ISSUE TITLE
gnn的一些疑问

### ISSUE THREAD

**Issue by searchlink** [PaddlePaddle/models#2768]
感谢提供paddle的实现， 目前在复现过程中遇到一些问题. 在文件夹`PaddleRec\gnn`
1.首先`reader.py中make_data()函数的定义`，关于label.append(e[1] - 1)中，为什么需要对label减1这样处理
2.关于`network.py中network()函数的定义`, 关于gnn多步传播的过程中，关于step>1的计算shape疑问。
```
 for i in range(step):
        pre_state = layers.reshape(x=pre_state, shape=[batch_size, -1, hidden_size])  
        state_in = layers.fc(input=pre_state, size=hidden_size, )
        state_out = layers.fc(input=pre_state, size=hidden_size, )
        state_adj_in = layers.matmul(adj_in, state_in)
        ...
        pre_state, _, _ = fluid.layers.gru_unit( input=gru_fc,
            hidden=layers.reshape(x=pre_state, shape=[-1, hidden_size]),
            size=3 * hidden_size)  
```
当 `step>1, i =0` , 此时`pre_state：[batch_size, uniq_max, hidden_size]`, 经过GRU之后的`pre_state`输出:  `[batch_size, hidden_size]`.
当i=1时， 此时`pre_state： [batch_size, 1, hidden_size]`，`adj_in: [batch_size, uniq_max, uniq_max]`. 此时`layers.matmul(adj_in, state_in)： [batch_size, uniq_max, uniq_max] * [batch_size, 1, hidden_size]`无法进行运算。麻烦帮忙解释下~多谢

**Comment by hutuxian** [comment]
感谢您的关注，很高兴为您回答问题：
1. 因为emb的第0维用于padding，不对应任何实际的item，所以网络里(network)的all_vocab是从1开始的。因为softmax操作里面的logits是从0开始的，为了能对应上，所以在做数据的时候把label减了个1。
2. 经过GRU之后的pre_state输出应该是[batch_size * uniq_max, h]维度的，然后在下一次循环的开始会reshape回[batch_size, uniq_max, h]这个维度。猜测您给出的shape是编译时维度，运行时的维度可以通过Print OP打印出来查看。

希望对您有帮助，欢迎其他问题与讨论，谢谢！

**Comment by zhiweilin** [comment]
label -1  可以一开始字典映射的时候从0开始，然后不用-1 吗
